monitoring, Adding a new alert KubemacpoolDown

RamLavi opened the pull request:
**What this PR does / why we need it**:
This PR Adds a new alert KubemacpoolDown
This alert will fire only when KMP is configured and for some reason the kubemacool-manager pod is not running.
The alert priority is set to critical, as if KMP is down, vms cannot be created/updated from the cluster.

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

RamLavi commented:
open issues:
- seems to be missing a service in order for the endpoint to be seen by the GUI
- unit tests
- e2e negative cases
- when removing CR the value stays 1. need to figure out how to zero this.

RamLavi commented:
manually checked on CRC

oshoval commented on pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:
```diff
@@ -286,6 +287,8 @@ func (r *ReconcileNetworkAddonsConfig) Reconcile(request reconcile.Request) (rec
 		return reconcile.Result{}, err
 	}
 
+	monitoring.TrackMonitoringComponents(&networkAddonsConfig.Spec, r.clusterInfo.MonitoringAvailable)
```
maybe should be after `deleteObjects` ? (actually same question is about `trackDeployedObjects` but not part of this PR)

https://github.com/kubevirt/cluster-network-addons-operator/pull/955/files#diff-463a5613afbb95588d747500e304c2d348c7fce8ec35117067322e7aac36a8dcR296

RamLavi replied:
I agree it's a more fitting place.. done

oshoval commented on pkg/monitoring/metrics.go:
```diff
@@ -26,10 +27,15 @@ var (
 			Name: "kubevirt_cnao_cr_ready",
 			Help: "Cnao CR Ready",
 		})
+	kmpDeployedGauge = prometheus.NewGauge(
```
kmp prefix is needed ? not sure

RamLavi replied:
what do you mean?

oshoval replied:
i confused something, nevermind, its fine

oshoval commented on pkg/monitoring/metrics.go:
```diff
@@ -26,10 +27,15 @@ var (
 
 func init() {
-	metrics.Registry.MustRegister(readyGauge)
+	metrics.Registry.MustRegister(readyGauge, kmpDeployedGauge)
```
maybe worth that all the gauges will be part of some slice and here we can register with `...` ?
just nice to have

RamLavi replied:
I prefer it this way. it's more explicit.

oshoval commented on pkg/monitoring/metrics.go:
```diff
@@ -40,6 +46,36 @@ func SetReadyGauge(isReady bool) {
+
+func trackKubemacpoolMonitoring(conf *cnao.NetworkAddonsConfigSpec) error {
+	isKubemacpoolDeployed := false
+	if conf.KubeMacPool != nil {
```
conf.KubeMacPool means its actually deployed ?
or just that it appeared on the manifest, and might be broken / just now reconcile started to deploy it ?

oshoval commented on pkg/monitoring/metrics.go:
```diff
@@ -40,6 +46,36 @@ func SetReadyGauge(isReady bool) {
+		return nil
+	}
+
+	err := trackKubemacpoolMonitoring(conf)
+	if err != nil {
+		return err
+	}
+
+	return nil
```
return trackKubemacpoolMonitoring(conf)

RamLavi replied:
This is intentional, I want this function to be a placeholder for future alerts.. similar to [Validate()](https://github.com/kubevirt/cluster-network-addons-operator/blob/main/pkg/network/network.go#L27)

perhaps I should do the `errs = append(...)` as well, what do you think?

oshoval replied:
ye maybe then worth to do the append, else its pretty creating infra for air-conditioner :)

RamLavi replied:
Done

oshoval reviewed (commented):
First pass, thanks

Can we add some info please about the alert on the PR description ?

oshoval commented on test/e2e/workflow/deployment_test.go:
```diff
@@ -281,25 +281,57 @@ var _ = Describe("NetworkAddonsConfig", func() {
+			expectedMetricValueMap map[string]string
+		}
+		DescribeTable("and checking scraped data",
+			func(p prometheusScrapeParams) {
```
can the func get 2 parameters instead of the struct ?
i don't see a reason to encapsulate those parameters into a struct

RamLavi replied:
I prefer it like this. 
helps keep parameter names understood , and makes thing neater imo.

oshoval commented on data/monitoring/prom-rule.yaml:
```diff
@@ -37,3 +37,15 @@ spec:
+          annotations:
+            summary: Kubemacpool is deployed by CNAO CR but kubemacpool pod is down.
+            runbook_url: http://kubevirt.io/monitoring/runbooks/KubemacpoolDown
+          expr: kubevirt_cnao_cr_kubemacpool_deployed_total == 1 and kubevirt_cnao_kubemacpool_manager_num_up_pods_total == 0
```
maybe due to a bug on CNAO, or just missing kmp on the CR
kubevirt_cnao_cr_kubemacpool_deployed / kubevirt_cnao_cr_kubemacpool_deployed_total will be zero, and then the alert won't raise, it might be fine, but just raising it to be sure

RamLavi replied:
the logic is that if KMP is not in the CR then KMP should *not* be deployed at all, hence the alert shouldn't fire (i.e. KMP is down but it's ok, and not something the sysAdmin should wake up in the middle of the night for).

oshoval commented on pkg/monitoring/metrics.go:
```diff
@@ -40,6 +46,36 @@ func SetReadyGauge(isReady bool) {
+		isKubemacpoolDeployed = true
+	}
+	setKubemacpoolDeployedGauge(isKubemacpoolDeployed)
+	return nil
```
not needed

RamLavi replied:
not sure what you mean here.. the function will return []error in general just so we have the option to return errors in the future.

oshoval replied:
i don't think we need to create infra for something that is not used right now, its a simple change that if someone need he can add a return value
atm, there is no other returns beside this one

RamLavi replied:
I disagree. Returning `[]error/error` from functions is generally a good practice. why not here?

oshoval replied:
The function has only one return statement and its always return nil
there is no good practice at all of returning the same thing no matter what the function is doing
i bet there are even static analyse tools that will warn / fail / optimize it

imo there should not be redundant code

RamLavi replied:
I tend to prefer adding the infra beforehand.
`TrackMonitoringComponents()` is the house for handling monitoring for all deployed components.
Each can potentially return an error, so `TrackMonitoringComponents()` collects the errors from them and flattens them to one error line if needed. 
Granted, currently there is only one: `trackKubemacpoolMonitoring()`, but I prefer that we lay out how we want to function to look like from the beginning, so that next time we want to add another monitoring (for knmstate for example) then that developer would only need to add the component's function.

We did the same when we added the `SpecialCleanUp()` [here](https://github.com/kubevirt/cluster-network-addons-operator/pull/270), and I think it's what we should do here as well.

oshoval commented on pkg/monitoring/metrics.go:
```diff
@@ -40,6 +46,36 @@ func SetReadyGauge(isReady bool) {
+	if conf.KubeMacPool != nil {
+		isKubemacpoolDeployed = true
+	}
+	setKubemacpoolDeployedGauge(isKubemacpoolDeployed)
```
if you like the style, the whole function can be one line 
setKubemacpoolDeployedGauge(conf.KubeMacPool != nil)

RamLavi replied:
I prefer the variable with the long name `isKubemacpoolDeployed` to make the develper's life easier to understand..

phoracek replied:
I am concerned about this hidden side effect. Could we either somehow lift this to an upper-level function or to rename this function to clearly show that it is changing the global? Maybe just use the setKubemacpoolDeployedGauge name directly? And if you could pass reference to the global to the function, it would get rid of the side effect completely

RamLavi replied:
I see, I'll move it upward, but not sure why I should pass the global. can you elaborate?

phoracek replied:
If you pass link to the global, you can unit test it. It would also become clear what is being "set". With it being taken from the global context, we cannot easily unit test it. It is unclear what this function works with unless you peek into it.

phoracek replied:
Sounds good

RamLavi commented:
@oshoval [fixed](https://github.com/kubevirt/cluster-network-addons-operator/compare/ad9259124901fbb570b3ad6d0ad227269e78b04e..85a2353e3eb7a322a82e1071024c3df33c50602f) and addressed comments

RamLavi commented:
> First pass, thanks
> 
> Can we add some info please about the alert on the PR description ?

added description in the PR

oshoval commented:
> This alert will fire only when KMP is configured and from some reason the kubemacool-manager pod is not running.

s/from/for/

oshoval reviewed (approved):
Thanks

RamLavi commented:
Added a function `ResetMonitoringComponents()` to reset the gouges when the CR is removed [here](https://github.com/kubevirt/cluster-network-addons-operator/compare/85a2353e3eb7a322a82e1071024c3df33c50602f..3516a4c9151adc5698758b7bfefa2c319483dad7)

RamLavi commented:
/test pull-e2e-cluster-network-addons-operator-br-marker-functests

oshoval commented on pkg/monitoring/metrics.go:
```diff
@@ -40,6 +47,42 @@ func SetReadyGauge(isReady bool) {
+func ResetMonitoringComponents() error {
+	SetReadyGauge(false)
+	setKubemacpoolDeployedGauge(false)
+	return nil
```
same deal here, i think return nil for a function that doesn't do anything else is totally unneeded,
its ok that a function won't return a value

maybe worth that the SetReadyGauge will be on other PR / commit, but i won't insist

RamLavi replied:
in this case I tend to agree there is no need for error in the function. will remove

RamLavi commented:
/retest

sradco commented:
/lgtm

phoracek commented on pkg/controller/networkaddonsconfig/networkaddonsconfig_controller.go:
```diff
@@ -297,6 +300,8 @@ func (r *ReconcileNetworkAddonsConfig) Reconcile(request reconcile.Request) (rec
 		return reconcile.Result{}, err
 	}
 
+	monitoring.TrackMonitoringComponents(&networkAddonsConfig.Spec, r.clusterInfo.MonitoringAvailable)
```
The condition MonitoringAvailable should be checked here. If it is not available, we should not invoke the module at all.  That way we encapsulate the helper module and isolate it from the dependencies. Would that make sense?

RamLavi replied:
I don't mind, but note we do it differently on render functions (see [here](https://github.com/kubevirt/cluster-network-addons-operator/blob/main/pkg/network/kubemacpool.go#L62-L64))
Do you think we should change it there as well? (in a different PR of course)

RamLavi replied:
Done

phoracek commented on data/monitoring/prom-rule.yaml:
```diff
@@ -37,3 +37,15 @@ spec:
             severity: warning
+        - expr: sum(up{namespace='{{ .Namespace }}', pod=~'kubemacpool-mac-controller-manager-.*'} or vector(0))
+          record: kubevirt_cnao_kubemacpool_manager_num_up_pods_total
+        - expr: sum(kubevirt_cnao_cr_kubemacpool_deployed{namespace='{{ .Namespace }}'} or vector(0))
```
Why do we need the sum here? Can't we just remove true or false from kubevirt_cnao_cr_kubemacpool_deployed ? We do not care about the number of deployed objects here, do we? Only about whether it is enabled or not

RamLavi replied:
the alert is using these 2 records `if(x==1 && y==0) for >5min then fire!` 
In order for the alert to work, the records needs to point to the same entry (x counts apples (arbitrary gauge) and y counts bananas (num of up kmp pods) )
After some playing with this with @sradco, we saw that `sum()` normalizes the records to be of same in "type" and so it works.

phoracek commented on pkg/monitoring/metrics.go:
```diff
@@ -40,6 +47,41 @@ func SetReadyGauge(isReady bool) {
+	}
+}
+
+func ResetMonitoringComponents() {
```
Are not these monitored components?

RamLavi replied:
They are! Thank you!!! 
finding the right wording brings me joy =)

RamLavi replied:
Done

phoracek commented on pkg/monitoring/metrics.go:
```diff
@@ -40,6 +47,41 @@ func SetReadyGauge(isReady bool) {
+}
+
+func trackKubemacpoolMonitoring(conf *cnao.NetworkAddonsConfigSpec) []error {
+	isKubemacpoolDeployed := false
```
Would this work?

```go
isKubemacpoolDeployed := conf.KubeMacPool != nil && conf.KubeMacPool
```

RamLavi replied:
like! Done

phoracek reviewed (approved):
/lgtm
/approve
